I1027 Add Contest Clock To WTI

clevengr opened the pull request:
### Description of what the PR does

Adds "Elapsed Time" and "Remaining Time" counters to the WTI header bar whenever a team is logged in.  

The counters are initialized with values read from the PC2 Server (via the WTI-API Server and its PC2 API connection) each time a team logs in.  The counters respond appropriately to changes in the "PC2 contest clock" state -- that is, when the Admin "Starts" the contest the counters begin counting (up/down respectively), and the counters stop (freeze) whenever the Admin "Stops" (pauses) the contest.  The counters also respond correctly to changes in the contest times; for example, if the Admin extends the contest (adds more time), then the "Remaining Time" counter automatically adjusts as appropriate.

Times are displayed in HH:MM:SS format, preceded by a number of days if the total time is greater than 24 hours.

The counters are driven by a JavaScript _setInterval()_ object which ticks at a rate of once per second.  To handle potential clock drift, the code automatically "resynchronizes" the on-screen counters with the current PC2 Server time, at a frequency defined by a constant in the WTI-UI `constants.ts` file (currently every 15 minutes).

A screenshot showing the clocks for a contest that was scheduled to run for 30 hours and has been running for just over an hour is shown here:

![image](https://github.com/user-attachments/assets/ebde809b-618c-4ac2-9440-a08d16e0d579)



### Issue which the PR addresses:

Fixes #1027, #1029

### Environment in which the PR was developed (OS,IDE, Java version, etc.)

- Windows 10 & 11
- Eclipse Versions 2021-12 (4.22.0) and 2020-12 (4.18.0)
- Java versions 1.8.0_381 and 11.0.16.1
- Chrome Version 131.0.6778.109
- MS Edge Version 131.0.2903.112 


### Precise steps for _testing_ the PR :

- Download and unzip the PR distribution.
- Edit samps/contests/sumithello/config/contest.yaml:  change from running:true to running:false,then save the updated file.
- Start a PC2 server using --load sumithello.
- Start a PC2 Admin; verify that the contest is NOT running and has NEVER BEEN STARTED (check the "Times" tab"; under "Started?" it should say "No").
- In the PC2 "projects" folder, unzip the WebTeamInterface .zip file.
- Go into the resulting WebTeamInterface folder, open a terminal command prompt, and start the WTI server with the command `./bin/pc2wti`.
- Open a browser, then hit F12 (or the browser's equivalent) to open the browser devtools; select the "Console" tab.
- Enter "http://localhost:8080" in the browser address bar.  Verify:
   - you are presented with the PC2 Login screen, and 
  - no ERRORS appear on the browser console (there will be lots of debug/tracing messages, but none should say "Error").
- Login as "team1/team1".  Verify:
   - you are presented with the PC2 "Runs" screen
   - the Header displays a field "Elapsed Time: 00:00:00 (no contest time has elapsed because the contest hasn't been started)
   - the Header displays a field "Remaining Time: xx:xx:xx", where xx:xx:xx is the number of hours/mins/secs defined as the
      length of the contest (for the default values in the "sumithello" contest, that should be "05:00:00")
      Note: if you change the Contest Length in the contest.yaml file BEFORE STARTING THE SERVER, and the value you set for
        contest length is greater than 24 hours, the "Remaining Time" field should also include a count of the number of DAYS remaining.
   - neither the Elapsed Time nor the Remaining Time values are changing (because the contest isn't running).
   - there are no ERRORS on the browser console (again, there will be lots of tracing/debug messages, but there should be none marked ERROR)
- In the browser, click on the Clarifications menu item; verify there are no ERROR messages in the browser console.
- Click on the Runs menu item, then click on the "Select Problem" dropdown in the "Filters" pane.  Verify that the contest problems are NOT shown.
- On the PC2 Admin, START the contest, then verify:
   - the Elapsed Time field on the browser header starts counting up at a rate of once per second;
   - the Remaining Time field on the browser header starts counting down at a rate of once per second;
   - clicking on the "Select Problem" dropdown in the "Filters" pane now shows the contest problems (Sumit and Hello).
   - there are no ERROR messages on the browser console.
- On the PC2 Admin, STOP ("pause") the contest, then verify:
   - both clock fields on the browser freeze;
   - the sum of the Elapsed time plus the Remaining Time on the browser header equals the "Contest Length" shown on the Admin "Length" field.
   - clicking on the "Select Problem" dropdown in the "Filters" pane still shows the contest problems.
   - there are no ERROR messages on the browser console.
- Note the "frozen" times in the browser window header, then on the PC2 Admin, START ("un-pause") the contest, then verify:
   - both clock fields start changing again, with Elapsed counting up and Remaining counting down, from the times they were "frozen" at when the contest was stopped.
   - there are no ERROR messages on the browser console.
- On the PC2 Admin, once again STOP the contest, then note the "Elapsed" and "Remaining" times which are now frozen on the browser header.
- Hit F5 in the browser window; verify that 
   - the Elapsed and Remaining Time clock values still have their same (frozen) values.
   - Clicking the "Select Problem" dropdown in the "Filters" pane still shows the contest problems.
- On the PC2 Admin, START the contest again; verify that 
   - the Elapsed and Remaining Time clocks resume counting correctly from their frozen values.
    - Clicking the "Select Problem" dropdown in the "Filters" pane still shows the contest problems.
- On the browser window, hit the "Logout" menu item; verify that you are returned to the PC2 Login screen.
- Login as another team; verify that the Elapsed and Remaining Time counters start at the expected values and are correctly changing  (the current "Elapsed" and "Remaining" times can be seen by hitting the "Refresh" button on the Admin "Times" tab).
- On the Admin, STOP the contest, then EXTEND the length of the contest by some amount (on the Admin "Times" screen, select Site1, click "Edit",   add some time to "Contest Length", and also add a corresponding amount of time to "Current Remaining Time".) Click "Update", then verify that  the "Remaining Time" on the team browser is properly updated.
- On the Admin, START the contest, then verify that the browser times begin properly changing again (and that "Remaining Time" correctly shows the updated value).
- MINIMIZE the team browser, wait a minute, then restore the browser.  Watch the "Elapsed Time" field as the browser is restored; it will start by displaying the elapsed time as it was AT THE MOMENT IT WAS MINIMIZED but should then (almost immediately) be updated to the correct elapsed time (which can be determined by hitting the "Refresh" button on the Admin "Times" window).  (This is happening because JavaScript timers don't run at the correct rate when the browser is minimized; 
a separate issue will be filed for this.)

Commit:
i1027: added contest timers to app-header files in "shared" module.

Commit:
i1027: added contest timers to app-header files in "shared" module.

Commit:
i1027: added ngx-countdown@v16.0 to package.json

Commit:
i1027: added ngx-countdown@v16.0 to package.json

Commit:
i1027: added various attempts to get HTML properly displaying clocks.

Commit:
i1027: added various attempts to get HTML properly displaying clocks.

Commit:
i1027: added setInterval object to gen timer ticks; added date func.

Commit:
i1027: added setInterval object to gen timer ticks; added date func.

Commit:
i1027: added attempts to access new ElapsedTimePipe from core module.

Commit:
i1027: added attempts to access new ElapsedTimePipe from core module.

Commit:
i1027: added attempts to access ElapsedTimePipe from shared module.

Commit:
i1027: added attempts to access ElapsedTimePipe from shared module.

Commit:
i1027: added new ElapsedTimePipe for formatting dates into dd:hh:mm:ss.

Commit:
i1027: added new ElapsedTimePipe for formatting dates into dd:hh:mm:ss.

clevengr committed:
i1027: copy 'constant.ts' into i1027 branch from PR871 branch.

clevengr committed:
i1027: added method getContestClock() to ContestServices (to both...

the abstract and concrete versions)

clevengr committed:
i1027: added method getContestClock() to ContestServices (to both...

the abstract and concrete versions)

clevengr committed:
i1027: added ContestClockModel to hold the WTI-API rep of ContestTime.

clevengr committed:
i1027: added ContestClockModel to hold the WTI-API rep of ContestTime.

clevengr committed:
i1027: added support for getting contest start time via the PC2 API.

clevengr committed:
i1027: added support for getting contest start time via the PC2 API.

clevengr committed:
i1027: add support for URL /contestclock to WTI-API server.

clevengr committed:
i1027: add support for URL /contestclock to WTI-API server.

clevengr committed:
i1027: fix minor typo in javadoc comment.

clevengr committed:
i1027: fix minor typo in javadoc comment.

clevengr committed:
i1027: add elapsedTimePipe to core/services -- an Angular "Pipe" that...

can be used to transform Date objects into strings displaying
day/hours/minutes/seconds.

clevengr committed:
i1027: add elapsedTimePipe to core/services -- an Angular "Pipe" that...

can be used to transform Date objects into strings displaying
day/hours/minutes/seconds.

clevengr committed:
i1027: clean up app-header HTML to display elapsed time in two ways:

- a count of the number of elapsed seconds (planned for removal), and
- a string showing days plus hh:mm:ss (the real plan).

clevengr committed:
i1027: clean up app-header HTML to display elapsed time in two ways:

- a count of the number of elapsed seconds (planned for removal), and
- a string showing days plus hh:mm:ss (the real plan).

clevengr committed:
i1027: minor doc (comments) added.

clevengr committed:
i1027: minor doc (comments) added.

clevengr committed:
i1027: added newly-required getContestClock() method to mock service.

clevengr committed:
i1027: added newly-required getContestClock() method to mock service.

clevengr committed:
i1027: added fetch of PC2 contest clock upon successful login.

clevengr committed:
i1027: added fetch of PC2 contest clock upon successful login.

clevengr committed:
i1027: removed no-longer-needed debugging; reformat code for readability

clevengr committed:
i1027: removed no-longer-needed debugging; reformat code for readability

clevengr committed:
i1027: rename time display pipe, rework to accept numSeconds

clevengr committed:
i1027: rename time display pipe, rework to accept numSeconds

clevengr committed:
i1027: add WTI-UI model for contest clock.

clevengr committed:
i1027: add WTI-UI model for contest clock.

clevengr committed:
i1027: added Timer class to enable/disable incr of display times.

clevengr committed:
i1027: added Timer class to enable/disable incr of display times.

clevengr committed:
i1027: remove no-longer-used ngx-countdown.

clevengr committed:
i1027: remove no-longer-used ngx-countdown.

clevengr committed:
i1027: include title-setting in all routes, matching PR871.

clevengr committed:
i1027: update for compatibility with PR for i871.

clevengr committed:
i1027: update for compatibility with PR for i871.

clevengr committed:
i1027: SharedModule: remove ngx-countdown; add ContestServices.

clevengr committed:
i1027: SharedModule: remove ngx-countdown; add ContestServices.

clevengr committed:
i1027: rename contestClock Subject to contestClockEvent; add debugging

clevengr committed:
i1027: rename contestClock Subject to contestClockEvent; add debugging

clevengr committed:
i1027: make some console debug output conditional on DEBUG_MODE

clevengr committed:
i1027: LoginPage: move ContestClock to ContestService; add doc and debug

clevengr committed:
i1027: LoginPage: move ContestClock to ContestService; add doc and debug

clevengr committed:
i1027: IWebSocketService: make root-injectable; add debug output.

clevengr committed:
i1027: IWebSocketService: make root-injectable; add debug output.

clevengr committed:
i1027: IContestService: make root-injectable; also:

- renamed contestClock to contestClockEvent;
- added support for unique instant ids;
- added abstract accessors for contest clock times;

clevengr committed:
i1027: IContestService: make root-injectable; also:

- renamed contestClock to contestClockEvent;
- added support for unique instant ids;
- added abstract accessors for contest clock times;

clevengr committed:
i1027: DisplayTimePipe: added debug output.

clevengr committed:
i1027: DisplayTimePipe: added debug output.

clevengr committed:
i1027: ContestTimerService: added error output when remainingTime<0

clevengr committed:
i1027: ContestTimerService: added error output when remainingTime<0

clevengr committed:
i1027: ContestService: make root-injectable; add ContestTimer; also:

- add support for updating ContestClock
- add debug output

clevengr committed:
i1027: ContestService: make root-injectable; add ContestTimer; also:

- add support for updating ContestClock
- add debug output

clevengr committed:
i1027: ContestMockService: add mock support for updating contest clock

clevengr committed:
i1027: ContestMockService: add mock support for updating contest clock

clevengr committed:
i1027: AppComponent: add support matching i871-F5-in-WTI.

clevengr committed:
i1027: AppComponent: add support matching i871-F5-in-WTI.

clevengr committed:
i1027: moved timer support out of AppHeader (into ContestService); also:

replaced use of ngx-countdown for remaining time with ContestService
timers.

clevengr committed:
i1027: moved timer support out of AppHeader (into ContestService); also:

replaced use of ngx-countdown for remaining time with ContestService
timers.

clevengr committed:
i1027: ContestTimerService:  added conditional debugging output.

clevengr committed:
i1027: ContestTimerService:  added conditional debugging output.

clevengr committed:
i1027: ContestService: added conditional debugging output.

clevengr committed:
i1027: ContestService: added conditional debugging output.

clevengr committed:
i1027: ContestService: return elapsed/remainging from Timer.

clevengr committed:
i1027: ContestService: return elapsed/remainging from Timer.

clevengr committed:
i1027: LoginPageComponent: added conditional debugging output.

clevengr committed:
i1027: LoginPageComponent: added conditional debugging output.

clevengr committed:
i1027: AppHeaderComponent: added conditional debugging output.

clevengr committed:
i1027: AppHeaderComponent: added conditional debugging output.

clevengr committed:
i1027: ContestClock:  added additional documentation.

clevengr committed:
i1027: ContestClock:  added additional documentation.

clevengr committed:
i1027: commented-out no-longer-needed debug output.

clevengr committed:
i1027: commented-out no-longer-needed debug output.

clevengr committed:
i1027: updated debug comments; added "todo".

clevengr committed:
i1027: updated debug comments; added "todo".

clevengr committed:
i1027: refactor updateContestClock() to be a ContestService method.

clevengr committed:
i1027: refactor updateContestClock() to be a ContestService method.

clevengr committed:
i1027: IWebsocketService: invoke updateContestClock() on 'clock' msg.

clevengr committed:
i1027: IWebsocketService: invoke updateContestClock() on 'clock' msg.

clevengr committed:
i1027: LoginPageComponent: add comments (no code changes).

clevengr committed:
i1027: LoginPageComponent: add comments (no code changes).

clevengr committed:
i127: AppHeaderComponent: remove debugging display of raw seconds.

clevengr committed:
i127: AppHeaderComponent: remove debugging display of raw seconds.

clevengr committed:
i1027: WTI-API.ContestController: fix copy/paste typo in response.

clevengr committed:
i1027: WTI-API.ContestController: fix copy/paste typo in response.

clevengr committed:
i1027: ContestTimer: set default remaining time to zero.  Note that...

...all logic paths SHOULD override this by getting the ACTUAL remaining
time from the PC2 server and updating the timer accordingly.

clevengr committed:
i1027: ContestTimer: set default remaining time to zero.  Note that...

...all logic paths SHOULD override this by getting the ACTUAL remaining
time from the PC2 server and updating the timer accordingly.

clevengr committed:
i1027: added comments (no code changes).

clevengr committed:
i1027: added comments (no code changes).

clevengr committed:
i1027: added toString() method.  This is important because...

...the class already defines a hashCode() method, and that hashCode()
method references the local variable "elementId" -- which is NULL!
The result is that anyone who called toString() (expecting to be able to
use the toString() inherited from Object) was getting a
NullPointerException (because Object's toString() invokes the hashCode()
method, which attempts to dereference elementId).

A separate question is "why does ContestTimeImplementation have a null
elementId?"   I'll file a separate issue for that..

clevengr committed:
i1027: added toString() method.  This is important because...

...the class already defines a hashCode() method, and that hashCode()
method references the local variable "elementId" -- which is NULL!
The result is that anyone who called toString() (expecting to be able to
use the toString() inherited from Object) was getting a
NullPointerException (because Object's toString() invokes the hashCode()
method, which attempts to dereference elementId).

A separate question is "why does ContestTimeImplementation have a null
elementId?"   I'll file a separate issue for that..

clevengr committed:
i1027: handle wallClockStartTime when contest hasn't been started; also,

added error logging to support debugging.

clevengr committed:
i1027: handle wallClockStartTime when contest hasn't been started; also,

added error logging to support debugging.

clevengr committed:
i1027: update contest clocks on F5 refresh/reload.

clevengr committed:
i1027: update contest clocks on F5 refresh/reload.

Commit:
i1027, i1029: update "contest is running" req for viewing clars & probs:

This commit removes the requirement that the contest be running in order
for the WTI-API /clarifications endpoint to return clars; now the only
requirement for returning clars is that the client "is-logged-in".

The commit also changes the requirement in the /problems endpoint:
previously the contest had to be RUNNING in order to see problems; now,
it just has to have been STARTED (i.e., there must have been greater
than zero "elapsed time".  This allows teams to see problems during a
pause.

Note that this commit is part of a PR for
https://github.com/pc2ccs/pc2v9/issues/1027, but it solves
https://github.com/pc2ccs/pc2v9/issues/1029 as well.

Commit:
i1027, i1029: update "contest is running" req for viewing clars & probs:

This commit removes the requirement that the contest be running in order
for the WTI-API /clarifications endpoint to return clars; now the only
requirement for returning clars is that the client "is-logged-in".

The commit also changes the requirement in the /problems endpoint:
previously the contest had to be RUNNING in order to see problems; now,
it just has to have been STARTED (i.e., there must have been greater
than zero "elapsed time".  This allows teams to see problems during a
pause.

Note that this commit is part of a PR for
https://github.com/pc2ccs/pc2v9/issues/1027, but it solves
https://github.com/pc2ccs/pc2v9/issues/1029 as well.

Commit:
i1027: add code to resync UI clocks with PC2 periodically.

Commit:
i1027: add code to resync UI clocks with PC2 periodically.

Commit:
i1027: constants.ts: set UI clock resync interval to every 15 mins.

Commit:
i1027: constants.ts: set UI clock resync interval to every 15 mins.

Commit:
i1027: source code cleanup (remove outdated comments, etc.)

Commit:
i1027: source code cleanup (remove outdated comments, etc.)

Commit:
i1027: added documentation.  No code changes.

Commit:
i1027: added documentation.  No code changes.

Commit:
i1027: use "start time" not elapsed secs to determine "started"; also:

added additional exception handling and updated HTTP error response
messages.

Commit:
i1027: use "start time" not elapsed secs to determine "started"; also:

added additional exception handling and updated HTTP error response
messages.

Commit:
i1027: allow problem display if contest started (not just if running).

Commit:
i1027: allow problem display if contest started (not just if running).

Commit:
i1027: pass ContestService to ContestTimer constructor; add "TODO" note.

Commit:
i1027: pass ContestService to ContestTimer constructor; add "TODO" note.

Commit:
i1027: Add code to handle time elapsed while browser is minimized; also:

added documentation

Commit:
i1027: Add code to handle time elapsed while browser is minimized; also:

added documentation

clevengr committed:
i1027: rename method to updateLocalContestClockFromServer()

clevengr committed:
i1027: rename method to updateLocalContestClockFromServer()

clevengr committed:
i1027: ContestTimer: always update "mostRecentTimerUpdate"; also:

apply renaming of updateContestClock() to
updateLocalContestClockFromServer().

clevengr committed:
i1027: ContestTimer: always update "mostRecentTimerUpdate"; also:

apply renaming of updateContestClock() to
updateLocalContestClockFromServer().

Commit:
i1027: removed multiple no-longer-needed debugging outputs.

Commit:
i1027: removed multiple no-longer-needed debugging outputs.

Commit:
i1027: added debug output.

Commit:
i1027: added debug output.

Commit:
i1027: added documentation (no code changes)

Commit:
i1027: added documentation (no code changes)

Commit:
i1027: update core.module.ts as part of rebase onto develop.

Commit:
i1027: fix conflict errors during rebase.

Commit:
i1027: fix merge conflicts:

Conflicts:
	projects/WTI-UI/src/app/modules/core/services/contest.service.ts
	projects/WTI-UI/src/app/modules/login/components/login-page/login-page.component.ts

kkarakas commented on projects/WTI-UI/src/app/modules/core/services/contest.service.ts:
```diff
@@ -74,5 +117,100 @@ export class ContestService extends IContestService {
+	 *  if the received data indicates the clock should be running it starts the ContestTimer (which then genrates a "clock tick"
+	 *  once per second to update the clock displays).
+	 */
+	installNewContestClock(data: any) {
```
Using any type, unless there is a good reason (such as I/O), is frowned upon.

kkarakas replied:
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#:~:text=%E2%9D%8C%20Don't%20use%20any,every%20usage%20of%20the%20variable.

clevengr replied:
Agreed; I'll fix it.

clevengr replied:
Fixed in next push.

kkarakas commented on projects/WTI-UI/src/app/modules/core/services/contest.service.ts:
```diff
@@ -74,5 +117,100 @@ export class ContestService extends IContestService {
+	 *  ContestTimer object.
+	 */
+	getElapsedSecs(): number {
+		let elapsedSecs = this.contestTimer.getElapsedSecs();	
```
Is there a reason this is a "let", it can be const. Or it can be returned directly.

clevengr replied:
I agree it should be const; I'll change it.  I prefer not to return expressions, for the reason that if the contents of the function needs to be modified/expanded later on, it could require making changes to the returned expression as well.  If someone misses this change while modifying the function it could lead to returning incorrect results. (I grant that since this is a simple one-line function it's not likely, but as a general principle I think it's safer to compute the return value within the function and then return the variable holding that return value.   Certainly, other people might disagree.

clevengr replied:
Fixed in next push.

kkarakas commented on projects/WTI-UI/src/app/modules/core/services/contest.service.ts:
```diff
@@ -74,5 +117,100 @@ export class ContestService extends IContestService {
+	 *  ContestTimer object.
+	 */
+	getRemainingSecs(): number {
+		let remainingSecs = this.contestTimer.getRemainingSecs();
```
Is there a reason this is a "let", it can be const. Or it can be returned directly.

clevengr replied:
Same reply as the previous comment; I'll change it to `const`.

clevengr replied:
Fixed in next push.

kkarakas commented on projects/WTI-UI/src/app/modules/core/services/contest.service.ts:
```diff
@@ -74,5 +117,100 @@ export class ContestService extends IContestService {
+		}
+		this.contestClock = newContestClock;
+		
+		//pull the values out of the updated clock
```
Similar comment on usage of "let"

clevengr replied:
And similar reply; I agree and I'll change them all to `const`.

clevengr replied:
Fixed in next push.

kkarakas commented on projects/WTI-UI/src/app/modules/login/components/login-page/login-page.component.ts:
```diff
@@ -124,17 +129,49 @@ export class LoginPageComponent implements OnInit, OnDestroy {
+						console.error ("LoginPageComponent.onSubmit() ContestClock subscription callback: unable to get ContestClock from PC2 API via ContestService!");
+					} else {						
+						//copy the data fields received from the PC2 Server (via the WTI-API) into the local ContestClock object
+						this.contestClock.isRunning = data.running ;
```
There is a lot of nested conditionals here, maybe we can use break to break out of conditions. This would improve readability.

clevengr replied:
I don't quite follow this comment.  As far as I can tell there are no nested conditionals which would allow using `break`.  What there ARE (or at least, WERE) are a lot of "debug if's".  I will remove all those debug statements; hopefully the logic will be clearer.

kkarakas replied:
I meant return sorry. Maybe the method can be terminated early if it is not going to lead anywhere?

clevengr replied:
I've removed the large number of "debug" statements, which makes the nested conditionals pretty easy to follow (as JohnB has noted elsewhere, the traditional JavaScript usage of "lamba functions" does complicate it a little bit).   As it now stands I don't really see where there's any place that the addition of "return" statements would be particularly helpful...

Let me know if you have a specific suggestion (short of rewriting the whole thing avoiding lambda expressions :) for improving it further.  If not, please mark the conversation "Resolved".

kkarakas commented on projects/WTI-UI/src/app/modules/shared/components/problem-selector/problem-selector.component.ts:
```diff
@@ -2,97 +2,162 @@ import { Component, forwardRef, OnInit, OnDestroy, Input } from '@angular/core';
+	 * and if wallClockStartTime is greater than zero it fetches the current problem list from the server and displays it in the
+	 * component; otherwise it sets the list of displayed problems to an empty list (array).
+	 */
+	private loadProblems(): void {
```
woah. Is it possible to simplify this method through having multiple methods breaks etc.

clevengr replied:
As with the previous comment, I don't see any way to use `break`s to address this.  As before, I think the problem was the large number of "debug" `if`s in the code, making it hard to see the structure.  I've eliminated all those debug statements; hopefully that makes the code more readable.  I'll welcome any suggestion for how to use `break`s to simplify it further.

johnbrvc replied:
I think the in-line lambda function is what clutters it up.  Folks love doing that in javascript too, and it makes it very hard to follow and read.  I guess it's the nature of the beast.

kkarakas replied:
I meant return sorry. Maybe the method can be terminated early if it is not going to lead anywhere?

clevengr replied:
As noted, I've removed the large number of "debug" statements; hopefully this makes the code clearer.  Also, as @johnbrvc noted, the use of JavaScript "lambda functions" makes it a little hard to read.  (I didn't write that code originally, and I'm reluctant to change it as part of this PR.  If you think it still needs to be changed, I suggest submitting a separate Issue for cleaning that up...)

Again, let me know if you have a specific suggestion (short of rewriting the whole thing avoiding lambda expressions :) for improving it further. If not, please mark the conversation "Resolved".

kkarakas reviewed (changes requested):
Did not run it on my machine yet, but these are my opinions. Looks like it would work well!

clevengr committed:
i1027: updated data types as requested in PR review.

clevengr committed:
i1027: change data types, remove debug prints for clarity.

clevengr committed:
i1027: fix data types, remove debug prints, minor reformatting.

kkarakas commented on projects/WTI-UI/src/app/modules/core/services/contestTimer.service.ts:
```diff
@@ -0,0 +1,139 @@
+   * ContestService updateLocalContestClockFromServer() method to resync the local contest clock with the server.
+   */
+  startTimer() {
+	if (this.intervalId) {
+      console.error("ContestTimerService.startTimer():  call to startTimer() when timer is already running");
+      return;
```
Tabbing doesn't seem to match on my screen here and other places. Is this intentional?

kkarakas replied:
Or is just on my screen?

clevengr replied:
If you are talking about the way indentation appears when you look at the code in Eclipse, that's a function of your Eclipse settings -- both global settings and the setting for the particular editor you're viewing it with.  At one point in time we (the early members of the PC2 project) had agreed upon a common set of indentation settings; perhaps yours don't match?

(If it's NOT in Eclipse that you're talking about, then where is it?)

johnbrvc replied:
It appears that there is an actual "TAB" character on the lines in question.  As @clevengr pointed out, if your Eclipse settings are "right" (by "right" I mean that they match ours), it will line up properly.   What IS interesting is, the code indenting appears to violate the indenting rules that we use in the java code.  The particular line in @kkarakas example has a TAB SPACE SPACE.  It seems to me, if TAB is 4 spaces, then the next level of indenting should be a total of 8 spaces (however you want to arrive at that: TAB TAB, 8x SPACE, etc.)  Not sure where the "+2" spaces comes from, unless that's standard for TS?  I attached a screenshot of my general editor properties.
<img width="545" alt="image" src="https://github.com/user-attachments/assets/d3fafdec-14de-49cf-9323-57e96834e95e" />

clevengr replied:
I'm not 100% positive, but I *think* the "default standard" for TypeScript IS "2 space indents".  I say this only because almost all the TS code we inherited from the students who did the original WTI project was indented by 2 spaces (on each indent).   I know that has caused some aggravation for me...  Not sure how we "fix" this.

johnbrvc replied:
Ok, that's fine, but what they did initially, it seems, is they "know" that TAB is 4 spaces, which is 2 "TS" indents, one keypress.  Ugh.  So, I guess just set your TAB to 4 spaces, and you should be ok.  The problem is, adding new code, you have to be aware of this and increase levels by 2 spaces...

kkarakas commented on projects/WTI-UI/src/app/modules/shared/components/app-header/app-header.component.scss:
```diff
@@ -14,6 +14,16 @@ header {
   .teamid-container {
     color: white;
     font-size: 1.5rem;
+    padding-left: 8px;
+    padding-right: 8px;
+  }
+  
+  /* Set the container holding the contest clock value(s) to medium, white font */
+  .clock-container {
+	    color: white;
+    font-size: 1.0rem;
+    padding-left: 8px;
+    padding-right: 8px;
```
I have some concerns about how these pixel paddings would look in different window sizes. Did you play around with different window size and different pixel counts? Perhaps we could use relative units such as em/ rem, vw, just like you use it for font-size. But this is not necessarily problematic, there are cases where pixels are better.

clevengr replied:
I did try it using various window sizes, and I didn't spot any problems.  Do you have a specific example of a platform and window size which shows a problem?

Regarding the "pixel paddings", I didn't create those; they came from a CSS file I got from @johnbrvc .   But I haven't observed any issues with it (I tested it in both Chrome and Edge, using various window sizes).  Again, do you have an example of a specific situation where a problem is demonstrated?

kkarakas replied:
Doesn't seem to be an issue.

clevengr replied:
Based on @kkarakas comment below ("Doesn't seem to be an issue.") I'm marking this Resolved.

kkarakas commented on projects/WTI-UI/src/app/modules/shared/components/app-header/app-header.component.ts:
```diff
@@ -14,13 +25,31 @@ export class AppHeaderComponent {
   get teamId(): string { 
     let acctId = this._authService.username; 
     let teamId = acctId.substr(4);
     return teamId;
```
This could be changed to const while we are at it.

clevengr replied:
Agreed; fixed in next push.

kkarakas commented on projects/WTI-UI/src/app/modules/shared/components/app-header/app-header.component.ts:
```diff
@@ -14,13 +25,31 @@ export class AppHeaderComponent {
 
-  constructor(private _authService: AuthService) { }
+  getElapsedSecs(): number {
+	let secs = this._contestService.getElapsedSecs() ;
+    return secs;
```
:)

clevengr replied:
Ageed; fixed in next push.

kkarakas reviewed (changes requested):
More opinions!

clevengr committed:
i1027: requested updates per PR review, including:

- use concrete data types rather than "any".
- remove debug statements.
- use "const" instead of "let" in variable declarations where
applicable.

kkarakas reviewed (commented):
Code looks good to me. I ran the branch and tested contest clock. Works as advertised. However it does not support  remaining time well if the value is negative. The screenshot attached happens 13 seconds after the contest ends. Perhaps it should be 0 or it should continue showing negative time. 
<img width="178" alt="Screenshot 2025-04-03 015536" src="https://github.com/user-attachments/assets/fe9dbac9-e4c5-4047-89c6-de8d2b08f32f" />

Commit:
i1027: update WTI-API ContestClockModel to match WTI-UI data fields.

Commit:
i1027: reduce WTI-UI "resync clocks" interval from 15 to 5 minutes.

Commit:
i1027: CI: update WTI-UI app-footer copyright date.

Commit:
i1027: CI: remove Runs screen "RunType" filtering (workaround for i1031)

Commit:
i1027: CI: replace Run Type with Run ID on runs screen (see Issue 1031)

clevengr committed:
i1027: correct mismatch clock data field names.

johnbrvc reviewed (approved):
I am not pretending to be a typescript expert, but I have been through code quite a bit in the past 4 or 5 days and I have a general understanding of how it works, or better yet, why it DIDN'T work but now does.  

I have tested the latest version and it appears to work as designed.
I will leave it up the TS experts to determine if there are any coding issues with the TS.  TBH, I spent most of the time reviewing the generated Javascript (main.js) when testing.

clevengr commented:
This PR should not be merged yet; I am still working on fixing the issue that @kkarakas pointed out -- negative "Remaining Time" is not handled properly.  I hope to get that fixed today...

clevengr committed:
i1027: stop "Remaining Time" countdown at zero.

clevengr committed:
i1027: Set Remaining Time font to red when contest has ended.

clevengr commented:
> Code looks good to me. I ran the branch and tested contest clock. Works as advertised. However it does not support remaining time well if the value is negative. The screenshot attached happens 13 seconds after the contest ends. Perhaps it should be 0 or it should continue showing negative time. <img alt="Screenshot 2025-04-03 015536" width="178" src="https://private-user-images.githubusercontent.com/60805192/429803613-fe9dbac9-e4c5-4047-89c6-de8d2b08f32f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2MTQ1ODAsIm5iZiI6MTc0NDYxNDI4MCwicGF0aCI6Ii82MDgwNTE5Mi80Mjk4MDM2MTMtZmU5ZGJhYzktZTRjNS00MDQ3LTg5YzYtZGU4ZDJiMDhmMzJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE0VDA3MDQ0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0NzE2ZDQ2MmNiMzUxYmZlYmIyNjQwMThiMTEwNDAxNTY4OTRmYzk3MDA1OWFhYzBlNjQ4NzA3YzQ3ZWYwNWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oqRiWVrzWUFnPOGRSO8Un1Swi8U43yE7P20oXnWB1lc">

Good catch; fixed in latest push.  I made it so that the "Remaining Time" stops changing when it hits zero, and also that it displays in red when it is zero.  The following screenshot shows a 5-hour contest that has run 17 seconds past the end; the "Elapsed" time is still counting but the "Remaining Time" is frozen at zero, in red.

![image](https://github.com/user-attachments/assets/72a7e8d1-75a7-4a6a-83b2-4e713f907e8a)

kkarakas commented on projects/WTI-UI/src/app/modules/core/services/contest.service.ts:
```diff
@@ -74,5 +96,90 @@ export class ContestService extends IContestService {
+	installNewContestClock(data: ContestClock) {
+		
+		//copy the data fields (received from the PC2 Server via the WTI-API) into a new ContestService ContestClock object
+		let newContestClock = new ContestClock();
```
This can be const.

clevengr replied:
Fixed in next push.

kkarakas commented on src/edu/csus/ecs/pc2/api/IContestClock.java:
```diff
@@ -1,10 +1,13 @@
 // Copyright (C) 1989-2019 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
```
copyright.

clevengr replied:
Fixed in next push.

kkarakas commented on src/edu/csus/ecs/pc2/api/implementation/ContestTimeImplementation.java:
```diff
@@ -1,6 +1,8 @@
 // Copyright (C) 1989-2019 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
```
copyright.

clevengr replied:
Fixed in next push.

kkarakas reviewed (commented):
These are my comments, I'll run the code tonight and see what happens.

clevengr committed:
i1027: PR requested changes.

Commit:
i1027: debug failure of F5 handler for remote server -- DO NOT MERGE

kkarakas reviewed (approved):
LGTM! I ran the program, issues I pointed out have been solved.
